hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java

```diff
@@ -625,11 +625,12 @@ HFileBlock unpack(HFileContext fileContext, FSReader reader) throws IOException
         : reader.getDefaultBlockDecodingContext();
       // Create a duplicated buffer without the header part.
       ByteBuff dup = this.buf.duplicate();
-      dup.position(this.headerSize());
+      int totalChecksumBytes = totalChecksumBytes();
+      dup.position(this.headerSize()).limit(dup.limit() - totalChecksumBytes);
       dup = dup.slice();
       // Decode the dup into unpacked#buf
-      ctx.prepareDecoding(unpacked.getOnDiskSizeWithoutHeader(),
-        unpacked.getUncompressedSizeWithoutHeader(), unpacked.getBufferWithoutHeader(true), dup);
+      ctx.prepareDecoding(unpacked.getOnDiskSizeWithoutHeader() - totalChecksumBytes,
```

Comment:
getOnDiskSizeWithoutHeader actually doesn't include the checksum bytes, so I don't think this change is correct.

We've been looking at this same area in https://issues.apache.org/jira/browse/HBASE-27053. My latest comments there have a deep dive on what's going on here, though more related to compression than encryption.

Reply:
actually, i was thinking of the uncompressed size, which doesn't include checksum. you're correct that on-disk size does.

Reply:
Yes, In the method prepareDecoding(), the uncompressedSizeWithoutHeader doesn't include checkSum bytes, the onDiskSizeWithoutHeader does include checkSum bytes. Because in the method getBufferWithoutHeader(boolean withChecksum),  we can see that the withCheckSum value is true.

Reply:
I want to know why `unpacked.getBufferWithoutHeader(true)` is set as true, Is this checkSum bytes used in somewhere?